wefe/debias/double_hard_debias.py

```diff
@@ -0,0 +1,490 @@
+            embeddings_dict[word] for word in self.target_words[0 : 2 * n_words]
+        ]
+        kmeans = KMeans(n_cluster).fit(embeddings)
+        y_pred = kmeans.predict(embeddings)
+        correct = [1 if item1 == item2 else 0 for (item1, item2) in zip(y_true, y_pred)]
+        alignment_score = sum(correct) / float(len(correct))
+        alignment_score = max(alignment_score, 1 - alignment_score)
+        return alignment_score
+
+    def fit(
+        self, model: WordEmbeddingModel, definitional_pairs: Sequence[Sequence[str]],
+    ) -> BaseDebias:
+        """Compute the bias direction and obtain the principal components of the entire
+           set of vectors.
+
+        Parameters
+        ----------
+        model : WordEmbeddingModel
+            The word embedding model to debias.
+        definitional_pairs : Sequence[Sequence[str]]
+            A sequence of string pairs that will be used to define the bias
+            direction. For example, for the case of gender debias, this list
+            could be [['woman', 'man'], ['girl', 'boy'], ['she', 'he'],
+            ['mother', 'father'], ...].
+
+        Returns
+        -------
+        BaseDebias
+            The debias method fitted.
+        """
+        self.definitional_pairs = definitional_pairs
+
+        self._check_sets_size(self.definitional_pairs, "definitional")
+
+        # -------------------------------------------------------------------
+        # Obtain the embedding of each definitional pairs.
+        if self.verbose:
+            print("Obtaining definitional pairs.")
+
+        self.definitional_pairs_embeddings = get_embeddings_from_sets(
+            model=model,
+            sets=definitional_pairs,
+            sets_name="definitional",
+            warn_lost_sets=self.verbose,
+            normalize=False,
+            verbose=self.verbose,
+        )
+
+        # -------------------------------------------------------------------
+        # Identify the bias subspace using the defining pairs.
+        if self.verbose:
+            print("Identifying the bias subspace.")
+        self.bias_direction = self._identify_bias_subspace(
+            self.definitional_pairs_embeddings, verbose=self.verbose
+        ).components_[0]
+
+        # -------------------------------------------------------------------
+        # Obtain embeddings' mean
+        self.embeddings_mean = self._calculate_embeddings_mean(model)
+
+        # -------------------------------------------------------------------
+        # Obtain the principal components of all vector in the model.
+        if self.verbose:
+            print("Obtaining principal components")
+        self.pca = self._principal_components(model)
+
+        return self
+
+    def transform(
+        self,
+        model: WordEmbeddingModel,
+        bias_representation: List[str],
+        ignore: List[str] = [],
+        copy: bool = True,
+        n_words: int = 1000,
+        n_components: int = 4,
```

Comment:
The idea of the transform is that it follows the definition inherited from BaseDebias:

```suggestion
    def transform(
        self,
        model: WordEmbeddingModel,
        target: Optional[List[str]] = None,
        ignore: Optional[List[str]] = None,
        copy: bool = True,
    ) -> WordEmbeddingModel:
```

However, here you add extra parameters. These extra parameters should be declared in the fit method, as they will determine the transformation to be executed.

```python
 fit(
    self, 
    model: WordEmbeddingModel, 
    definitional_pairs: Sequence[Sequence[str]],
    bias_representation: List[str],
    n_words: int = 1000,
    n_components: int = 4,
) -> BaseDebias:
```

Then, if you have intermediate results derived from those parameters in fit, simply store them in self, so that they are waiting to be used in the transform method.

However, I am not sure how to implement the target in this case.

Reply:
Moved the parameters from transform and implemented target